OffscreenCanvas::transferToImageBitmap doesn't enforce size restrictions

mattwoodrow opened the pull request:
#### e4c57c810933fe6b9d74013303d92b3484b62f33
<pre>
OffscreenCanvas::transferToImageBitmap doesn&apos;t enforce size restrictions
<a href="https://bugs.webkit.org/show_bug.cgi?id=251130">https://bugs.webkit.org/show_bug.cgi?id=251130</a>

Reviewed by Chris Lord.

We share code with HTMLCanvasElement via CanvasBase for allocating the backing store, with max size restrictions.

transferToImageBitmap can sometimes allocate a new buffer (if there isn&apos;t one already), and it currently isn&apos;t using the same code, so doesn&apos;t have the same restrictions.

This makes CanvasBase::createImageBuffer return the allocated buffer so that we can use it for transferToImageBitmap and get identical behaviour between code paths.

* LayoutTests/platform/ios/fast/canvas/offscreen-giant-transfer-to-imagebitmap-expected.txt: Added.
* Source/WebCore/html/CanvasBase.cpp:
(WebCore::CanvasBase::buffer const):
(WebCore::CanvasBase::allocateImageBuffer const):
(WebCore::CanvasBase::createImageBuffer const): Deleted.
* Source/WebCore/html/CanvasBase.h:
(WebCore::CanvasBase::createImageBuffer const):
* Source/WebCore/html/HTMLCanvasElement.cpp:
(WebCore::HTMLCanvasElement::createImageBuffer const):
* Source/WebCore/html/HTMLCanvasElement.h:
* Source/WebCore/html/OffscreenCanvas.cpp:
(WebCore::OffscreenCanvas::transferToImageBitmap):
(WebCore::OffscreenCanvas::createImageBuffer const):
* Source/WebCore/html/OffscreenCanvas.h:

Canonical link: <a href="https://commits.webkit.org/260035@main">https://commits.webkit.org/260035@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/342fdd49a77fd6ec0977714019d53824e59e30a1

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/105390 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/14459 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/38263 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/114646 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/114530 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/109293 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/15648 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/5396 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/97694 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/111148 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/12092 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/95074 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/39578 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/93958 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/26709 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/81262 "Found 1 new API test failure: /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/text/state-changed (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/7773 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/28068 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/7893 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/4662 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/13916 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/47613 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6930 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/9684 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

Cwiiis commented on Source/WebCore/html/OffscreenCanvas.cpp:
```diff
@@ -524,10 +524,10 @@ void OffscreenCanvas::scheduleCommitToPlaceholderCanvas()
-void OffscreenCanvas::createImageBuffer() const
+RefPtr<ImageBuffer> OffscreenCanvas::createImageBuffer() const
 {
     m_hasCreatedImageBuffer = true;
```
Now that createImageBuffer doesn't associate a buffer with this OffscreenCanvas, it would make more sense to have an `m_hasSetImageBuffer` member instead of `m_hasCreatedImageBuffer` and to set this in setImageBuffer. There's the situation now (I think) where you could call `transferToImageBitmap` before a buffer is created, then `m_hasCreatedImageBuffer` will be set to `true`, but no image buffer would have been set.

Cwiiis commented on Source/WebCore/html/HTMLCanvasElement.cpp:
```diff
@@ -923,8 +923,10 @@ void HTMLCanvasElement::createImageBuffer() const
 
-    if (m_context && buffer() && buffer()->prefersPreparationForDisplay())
+    if (m_context && buffer && buffer->prefersPreparationForDisplay())
         const_cast<HTMLCanvasElement*>(this)->addObserver(document());
```
Similar to the below comment, it feels odd to me that we'd be adding the observer here and not in `setImageBuffer` - if `createImageBuffer` is now a utility function to create an image buffer and do the necessary checks and such any time a buffer is needed, it doesn't make sense to me that this logic would be here, before the created buffer is actually associated with the OffscreenCanvas.

Cwiiis reviewed (changes requested):
My general concern with this patch is that it's separating `createImageBuffer` into separate create and set steps, but there's logic in the create step that should be in the set step. I may have missed something and am open to being convinced otherwise though!

I guess it goes without saying that also there are some test expectations that need updating, but I've said it anyway :)

Cwiiis commented on Source/WebCore/html/OffscreenCanvas.cpp:
```diff
@@ -535,7 +535,7 @@ void OffscreenCanvas::scheduleCommitToPlaceholderCanvas()
 void OffscreenCanvas::createImageBuffer() const
```
I still have some confusion as to the relationship between allocate, set and create. Right now, it's possible to set a buffer and have `m_hasCreatedImageBuffer` be `false` - presumably this is intentional because we have this flag in the first place when you could just do a null check on the buffer pointer... But then, maybe the flag exists for descendants that aren't backed by a nullable type? It feels like this needs some straightening out to me.

The more obvious thing to me would be to remove `createImageBuffer` and just have `allocateImageBuffer` and `setImageBuffer`, and to override the virtual `hasCreatedImageBuffer` with just a null check on the image buffer pointer (this could then also be renamed to the shorter `hasImageBuffer`).

Failing that (which, in fairness, would be a much bigger change), I would prefer rather than peppering `m_hasCreatedImageBuffer = true` throughout the file, that for the one situation where we want it to be `false` with a set buffer to be singled out instead. It feels to me like it would be too easy right now to mistakenly set the buffer and forget to set the 'created' flag. It'd be great if it was all a bit more tightly coupled.

mattwoodrow replied:
It indeed is very confusing, I did some digging to try understand it a bit better.

It looks like `m_hasCreatedImageBuffer` is defined separately for `HTMLCanvasElement` and `OffscreenCanvas`, but both are used to implement the virtual `CanvasBase::hasCreatedImageBuffer`.

The `HTMLCanvasElement` has this comment: "m_hasCreatedImageBuffer means we tried to malloc the buffer. We didn't necessarily get it."

The most obvious use case I can see is the implementation of `CanvasBase::buffer()`, where we want to lazily create a buffer if we haven't yet, but we also don't want to keep doing that every time if it's failing to allocate.

For OffscreenCanvas as it is, I think we need `m_hasCreatedImageBuffer` to be false initially, until `createImageBuffer` (or `setImageBufferAndMarkDirty`) is called, and then true (even if alloc failed) until something consumes the buffer.

This might be better by making the underlying buffer type `std::optional<RefPtr<ImageBuffer>>` so that we can represent all three states without a separate variable that can be forgotten about. I think that'd be better as a separate change, but I can do that as a followup.

Cwiiis reviewed (changes requested):
This is better, but still a bit more error-prone than I'd like, personally. It could still be that I'm missing some of the bigger picture as it's been a while since I was in the canvas code. Would definitely appreciate some clarification if you think I've missed/misunderstood something.

Cwiiis reviewed (approved):
Given the previous comments and assuming the follow-up is being tracked, this sounds/looks good to me.

An interim before a more comprehensive change may be to just rename `m_hasCreatedImageBuffer` to `m_hasAttemptedBufferAllocation` or something like that.

mattwoodrow committed:
OffscreenCanvas::transferToImageBitmap doesn't enforce size restrictions
https://bugs.webkit.org/show_bug.cgi?id=251130

Reviewed by Chris Lord.

We share code with HTMLCanvasElement via CanvasBase for allocating the backing store, with max size restrictions.

transferToImageBitmap can sometimes allocate a new buffer (if there isn't one already), and it currently isn't using the same code, so doesn't have the same restrictions.

This makes CanvasBase::createImageBuffer return the allocated buffer so that we can use it for transferToImageBitmap and get identical behaviour between code paths.

* LayoutTests/platform/ios/fast/canvas/offscreen-giant-transfer-to-imagebitmap-expected.txt: Added.
* Source/WebCore/html/CanvasBase.cpp:
(WebCore::CanvasBase::buffer const):
(WebCore::CanvasBase::allocateImageBuffer const):
(WebCore::CanvasBase::createImageBuffer const): Deleted.
* Source/WebCore/html/CanvasBase.h:
(WebCore::CanvasBase::createImageBuffer const):
* Source/WebCore/html/HTMLCanvasElement.cpp:
(WebCore::HTMLCanvasElement::createImageBuffer const):
* Source/WebCore/html/HTMLCanvasElement.h:
* Source/WebCore/html/OffscreenCanvas.cpp:
(WebCore::OffscreenCanvas::transferToImageBitmap):
(WebCore::OffscreenCanvas::createImageBuffer const):
* Source/WebCore/html/OffscreenCanvas.h:

Canonical link: https://commits.webkit.org/260035@main
